### PR TITLE
[ RNN ] Backwarding Implementation @open sesame 04/01 11:12

### PR DESCRIPTION
**Pull request by jijoongmoon**
This commit includes calculate gradient and derivatives of RNN Layer.

Resolves:

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: jijoong.moon <jijoong.moon@samsung.com>

### CODE REVIEW

File: nntrainer/layers/rnn.cpp [hunk 1]
```diff
@@ -163,11 +164,68 @@ void RNNLayer::copy(std::shared_ptr<Layer> l) {
...
+    Tensor xs_t = input_.getBatchSlice(b, 1);
+    Tensor hs_t = hidden_.getBatchSlice(b, 1);
+    dh_nx.setZero();
+
+    Tensor dh;
+    Tensor xs;
+    Tensor hs_prev;
+    Tensor hs;
+
+    for (unsigned int t = deriv_t.height(); t-- > 0;) {
```
Comment (zhoonit) [inline nntrainer/layers/rnn.cpp:201]:
```suggestion
    for (int t = deriv_t.height(); t-- > 0;) {
```
It seems like t might go underflow

File: nntrainer/layers/rnn.cpp [hunk 2]
```diff
@@ -163,11 +164,68 @@ void RNNLayer::copy(std::shared_ptr<Layer> l) {
 }
 
 void RNNLayer::calcDerivative() {
-  //  NYI
+  Tensor &derivative_ = net_hidden[0]->getGradientRef();
+  Tensor &weight =
+    weightAt(static_cast<int>(RNNParams::weight_xh)).getVariableRef();
+  Tensor &ret_ = net_input[0]->getGradientRef();
+
+  ret_ = derivative_.dot(weight, ret_, false, true);
```
Comment (zhoonit) [inline nntrainer/layers/rnn.cpp:172]: Minor but this removes extra copy.

```suggestion
  derivative_.dot(weight, ret_, false, true);
```

File: nntrainer/layers/rnn.cpp [hunk 3]
```diff
@@ -163,11 +164,68 @@ void RNNLayer::copy(std::shared_ptr<Layer> l) {
...
+      } else {
+        hs_prev = hs_t.getSharedDataTensor(TensorDim(1, 1, 1, hs_t.width()),
+                                           (t - 1) * hs_t.width());
+      }
+
+      if (t < deriv_t.height() - 1) {
+        dh.add_i(dh_nx);
+      }
+
+      hs = acti_func.run_prime_fn(hs, dh, dh);
```
Comment (zhoonit) [inline nntrainer/layers/rnn.cpp:220]: Same.
```suggestion
      acti_func.run_prime_fn(hs, dh, dh);
```

File: nntrainer/layers/rnn.cpp (hunk 1 above)
Comment (jijoongmoon) [reply nntrainer/layers/rnn.cpp:201]: I don't think so. t never be negative.

File: nntrainer/layers/rnn.cpp (hunk 2 above)
Comment (jijoongmoon) [reply nntrainer/layers/rnn.cpp:172]: done.

File: nntrainer/layers/rnn.cpp (hunk 3 above)
Comment (jijoongmoon) [reply nntrainer/layers/rnn.cpp:220]: done.

File: nntrainer/layers/rnn.cpp (hunk 1 above)
Comment (zhoonit) [reply nntrainer/layers/rnn.cpp:201]: Right, I mistook.

**Review by zhoonit — APPROVED** [review]
LGTM

File: nntrainer/layers/rnn.cpp [hunk 4]
```diff
@@ -163,11 +164,68 @@ void RNNLayer::copy(std::shared_ptr<Layer> l) {
...
+
+      if (t < deriv_t.height() - 1) {
+        dh.add_i(dh_nx);
+      }
+
+      acti_func.run_prime_fn(hs, dh, dh);
+      dh.multiply_i(hs);
+      djdb_h.add_i(dh);
+
+      djdw_x.add_i(xs.dot(dh, true, false));
```
Comment (kparichay) [inline nntrainer/layers/rnn.cpp:224]: how about set replace add_i with providing the last parameter `beta` as 1.0 and initialize the djdw_x and djdw_h as zero?

File: nntrainer/layers/rnn.cpp [hunk 5]
```diff
@@ -163,11 +164,68 @@ void RNNLayer::copy(std::shared_ptr<Layer> l) {
...
+      if (t == 0) {
+        hs_prev = Tensor(TensorDim(1, 1, 1, hs_t.width()));
+        hs_prev.setZero();
+      } else {
+        hs_prev = hs_t.getSharedDataTensor(TensorDim(1, 1, 1, hs_t.width()),
+                                           (t - 1) * hs_t.width());
+      }
+
+      if (t < deriv_t.height() - 1) {
+        dh.add_i(dh_nx);
```
[the remark this answers is not in the document]
Comment (jijoongmoon) [reply nntrainer/layers/rnn.cpp:217]: yes. It is.

File: test/unittest/unittest_nntrainer_layers.cpp [hunk 6]
```diff
@@ -2351,24 +2351,39 @@ TEST_F(nntrainer_RNNLayer, initialize_01_p) {
...
+  allocateMemory();
   EXPECT_NO_THROW(layer.forwarding_with_val({input}, {}, false));
+
+  nntrainer::Tensor derivatives(2, 1, 3, 3);
+  derivatives.setValue(1.0);
+
+  setOptimizer(nntrainer::OptType::SGD, "learning_rate=1.0");
+
+  nntrainer::Tensor result;
+  EXPECT_NO_THROW(result = *layer.backwarding_with_val(
```
[the remark this answers is not in the document]
Comment (jijoongmoon) [reply test/unittest/unittest_nntrainer_layers.cpp:2385]: I will do later pr.

File: nntrainer/layers/rnn.cpp (hunk 4 above)
Comment (jijoongmoon) [reply nntrainer/layers/rnn.cpp:224]: right. I forget to set the gradients zero. I will add. 
>> add_i with providing the last parameter beta as 1.0
so.. you mean A= beta*A + B and if beta is 0.0 the A is zero??

Comment (kparichay) [reply nntrainer/layers/rnn.cpp:228]: > so.. you mean A= beta*A + B and if beta is 0.0 the A is zero??

right. If beta is 0.0, then the original value of A will not affect the result.

File: test/unittest/unittest_nntrainer_layers.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (kparichay) [reply test/unittest/unittest_nntrainer_layers.cpp:2385]: ok.

File: nntrainer/layers/rnn.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (kparichay) [reply nntrainer/layers/rnn.cpp:221]: There are two scenarios:
1. Hidden state is reset to zero every iteration - in this case, hidden state can be update in backwarding as it does not affect the next iteration.
2. Hidden state is NOT reset to zero every iteration - in this case, hidden state must retain its value for the next iteration.

In the forwarding, as per my understanding, it looks like scenario 2. Please let know, if I missed some part.

File: nntrainer/layers/rnn.cpp (hunk 4 above)
Comment (jijoongmoon) [reply nntrainer/layers/rnn.cpp:228]: It's good. I will update.

File: nntrainer/layers/rnn.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (jijoongmoon) [reply nntrainer/layers/rnn.cpp:221]: Thanks. We made offiline discussion.

**Review by kparichay — APPROVED** [review]
LGTM
